src/client/Microsoft.Identity.Client/Extensibility/AbstractConfidentialClientAcquireTokenParameterBuilderExtension.cs

```diff
@@ -0,0 +1,32 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Identity.Client.Extensibility
+{
+    /// <summary>
+    /// Extensions for <see cref="AcquireTokenForClientParameterBuilder"/>
+    /// </summary>
+    public static class AbstractConfidentialClientAcquireTokenParameterBuilderExtension
+    {
+        /// <summary>
+        /// Overrides the client credentials parameters (e.g. client_assertion) and optionally ties the 
+        /// resulting token to a key id, similar to POP tokens.
+        /// </summary>
+        /// <param name="clientAssertionOverride">A lambda which receives a string representing the value of the token endpoint and which needs to return key value pairs to be added to the POST payload</param>
+        /// <param name="builder">Builder to chain config options to</param>
+        /// <returns>The builder</returns>
+        public static AbstractConfidentialClientAcquireTokenParameterBuilder<T> WithClientAssertionOverride<T>
+            (this AbstractConfidentialClientAcquireTokenParameterBuilder<T> builder,
+            Func<string, IReadOnlyList<KeyValuePair<string, string>>> clientAssertionOverride)
```

Comment:
Potential changes: 

- make async
- move to App config.

Reply:
In mechanism2, you mean requestBuilder for the 2nd snippet, @bgavrilMS ?

Reply:
No, in mechanism2 I mean a lambda. The lambda is invoked on a per request basis.

Reply:
The lambda mechanism is too messy. I tried and the user code (in the integration test) is too complex.

Reply:
For me, they seem to be more or less the same. You always have the option of setting the func equal to a defined method instead of lambda and then it just becomes another parameter. 

Also, we already have 
`ConfidentialClientApplicationBuilder WithClientAssertion(string signedClientAssertion)`
On the app builder so it may seem confusing if someone is already used to this being on the app builder but suddenly the new override is moved to the request builder. Also, the client assertions represent application credentials and we tend to put these in the application builder.

Reply:
After playing around with this some more, it is clear to me that the app developer wanting to keep the CCA object as singleton, which is important for perf, needs to pass in the cert at request level. Having it at app level makes the programming much more complicated and prone to threading issues.